daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala

```diff
@@ -193,13 +193,71 @@ object SValue {
       discard[Int](`SMap Ordering`.compare(k, k))
     }
 
-    def apply(isTextMap: Boolean, entries: Iterator[(SValue, SValue)]): SMap = {
+    /** Build an SMap from an indexed sequence of SValue key/value pairs.
+      *
+      * SValue keys are assumed to be in ascending order - hence the SMap's TreeMap will be built in time O(n) using a
+      * sorted map specialisation.
+      */
+    def apply(
+        isTextMap: Boolean,
+        entries: IndexedSeqView[(SValue, SValue)],
+    ): SMap = {
+      require(
+        entries
+          .foldLeft[(Boolean, Option[SValue])]((true, None)) {
+            case ((result, previousKey), (currentKey, _)) =>
+              (result && previousKey.forall(`SMap Ordering`.lteq(_, currentKey)), Some(currentKey))
+          }
+          ._1,
+        "The entries are not in descending order",
+      )
+
+      val sortedEntryMap: SortedMap[SValue, SValue] = new SortedMap[SValue, SValue] {
+        private[this] val encapsulatedSortedMap = SortedMap.from(entries)
+
+        override def iterator: Iterator[(SValue, SValue)] = entries.iterator
+
+        override def size: Int = entries.size
+
+        override def updated[V1 >: SValue](key: SValue, value: V1): SortedMap[SValue, V1] =
+          encapsulatedSortedMap.updated(key, value)
+
+        override def removed(key: SValue): SortedMap[SValue, SValue] =
+          encapsulatedSortedMap.removed(key)
+
+        override def iteratorFrom(start: SValue): Iterator[(SValue, SValue)] =
+          encapsulatedSortedMap.iteratorFrom(start)
+
+        override def keysIteratorFrom(start: SValue): Iterator[SValue] =
+          encapsulatedSortedMap.keysIteratorFrom(start)
+
+        override def ordering: Ordering[SValue] = `SMap Ordering`
+
+        override def rangeImpl(
+            from: Option[SValue],
+            until: Option[SValue],
+        ): SortedMap[SValue, SValue] = encapsulatedSortedMap.rangeImpl(from, until)
+
+        override def get(key: SValue): Option[SValue] = encapsulatedSortedMap.get(key)
```

Comment:
@remy rather than implement these overrides with `???` I've gone for a more meaningful implementation (hoping that potential changes to Scala's `TreeMap` have a limited impact). If you would prefer a `???` implementation here (with the potential risk of unimplemented exceptions being thrown in the future) then I'm happy to change this.

Reply:
:+1:

Reply:
wrong @remy

Reply:
Apologies for that